### PR TITLE
Correctly seed root direntry for ZipArchives

### LINKED ISSUE #11754: scala.reflect.io.ZipArchive#ensureDir infinitely recurses in some cases in 2.13.1
It looks like this class was modified in 2.13.1 to use a "path" representation for entries always, instead of going back and forth between "dotted" and "path" representations (see scala/scala#8356). The `ensureDir` function depends on having a "root" entry already in the `dirs` map which is passed in - otherwise, it will never terminate, as the `dirName` function which it calls has a fixed point at `"/"`.

With the "path" representation, the root is always `"/"`, whereas it seems to have been `""` in some cases previously. This means that every time we will call this `ensureDir` functions, we need to make sure that we pass it a map already containing an entry for `"/"`. [There seems to have already been an effort](https://github.com/scala/scala/commit/0c7170b2abc696c55ec596cd715b3454db02e344#diff-a1226d4dc1d8e8e5d47949335b81a607R189) to fix some of these cases. Unfortunately, not all were caught.

In our case, we are reflectively compiling scala code at runtime, and are getting a `StackOverflowError` in this method when resolving one package (`javax.swing`). I have traced this to [this line](https://github.com/scala/scala/blob/59dec6114da327a1c79e3cc1571f4173ecb6e829/src/reflect/scala/reflect/io/ZipArchive.scala#L325), which I believe should use `"/"` as the key, instead of `""`. I think there is a similar bug at line 250 in that file, in `URLZipArchive`.

### PR DESCRIPTION
**Pull request by som-snytt**
As diagnosed.

Fixes scala/bug#11754

### CODE REVIEW

**Comment by som-snytt** [comment]
MiMa objects to my package private constant value definition for slash. Whitelisting is so tiresome, I may just revert that. Or maybe try the exclusion. Would be nice to have a `scala-sbt-fix` task to automate that!

File: test/junit/scala/tools/nsc/backend/jvm/BytecodeTest.scala [hunk 1]
```diff
@@ -307,6 +307,7 @@ class BytecodeTest extends BytecodeTesting {
     ), fs)
     val assignedInConstr = getMethod(k, "<init>").instructions.filter(_.opcode == Opcodes.PUTFIELD) map {
       case f: Field => f.name
+      case _ => ???  // @unchecked
```
Comment (lrytz) [inline test/junit/scala/tools/nsc/backend/jvm/BytecodeTest.scala:310]: yeah, `@unchecked` would be cleaner

**Review by retronym — APPROVED** [review]
Thank you.

**Comment by som-snytt** [comment]
Rebase, squash, resolve hot lines of mima exclusions.

File: src/reflect/scala/reflect/io/ZipArchive.scala [hunk 2]
```diff
@@ -245,9 +246,10 @@ final class FileZipArchive(file: JFile, release: Option[String]) extends ZipArch
 /** ''Note:  This library is considered experimental and should not be used unless you know what you are doing.'' */
 final class URLZipArchive(val url: URL) extends ZipArchive(null) {
   def iterator: Iterator[Entry] = {
-    val root     = new DirEntry("/")
+    val root     = new DirEntry(RootEntry)
     val dirs     = new java.util.HashMap[String, DirEntry]()
-    dirs.put("", root)
+    dirs.put(RootEntry, root)
+    val ss       = Streamable.bytes(input)
```
Comment (som-snytt) [inline src/reflect/scala/reflect/io/ZipArchive.scala]: debug cruft

**Comment by SethTisue** [comment]
@som-snytt this PR seems to be causing test failures on Windows: https://scala-ci.typesafe.com/view/scala-2.13.x/job/scala-2.13.x-integrate-windows/1157/ and all subsequent runs as well

```
[error] Failed tests:
[error] 	scala.reflect.io.ZipArchiveTest
[error] 	scala.tools.nsc.classpath.ZipAndJarFileLookupFactoryTest
```

**Comment by som-snytt** [comment]
@SethTisue thanks for the heads up. I don't know why I didn't run the test suite on a windows box a long time ago. I'll take a look.

**Comment by som-snytt** [comment]
Also, this profile image might be a little too close to :poop:

**Comment by som-snytt** [comment]
Possible fix at https://github.com/scala/scala/pull/8539

**Pull request merged**